packages/block-editor/src/components/block-list/use-block-props/use-block-nodes.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * External dependencies
+ */
+import { omit } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useContext } from '@wordpress/element';
+import { useRefEffect } from '@wordpress/compose';
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { SetBlockNodes } from '../';
+import { store as blockEditorStore } from '../../../store';
+
+export function useBlockNodes( clientId ) {
+	const setBlockNodes = useContext( SetBlockNodes );
+	// Provide the selected node, or the first and last nodes of a multi-
+	// selection, so it can be used to position the contextual block toolbar.
+	// We only provide what is necessary, and remove the nodes again when they
+	// are no longer selected.
+	const isNodeNeeded = useSelect(
```

Comment:
Does this check have any impact on performance? I mean if not, we could  say that the BlockNode context make all block nodes available.

Reply:
It does because it triggers a state change for every addition. Imagine loading a post: each mounted block will trigger another state change setting the node.

Perhaps we need to rethink this part a bit. Can we add ref callbacks only when needed? Maybe a filter on the block ref so you can add one based on the block ID?